muspinsim/experiment.py

```diff
@@ -205,16 +206,33 @@ def rho0(self):
 
     @property
     def Hz(self):
-
         if self._Hz is None:
-            B = self._B
-            g = self._system.gammas
-            Bg = B[None, :] * g[:, None]
+            # Compute Zeeman Hamiltonian contribution
+            if not self._config.celio_k:
+                B = self._B
+                g = self._system.gammas
+                Bg = B[None, :] * g[:, None]
 
-            Hz_sp_list = (self._single_spinops * Bg).flatten().tolist()
-            Hz = np.sum(Hz_sp_list)
+                Hz_sp_list = (self._single_spinops * Bg).flatten().tolist()
+                Hz = np.sum(Hz_sp_list)
 
-            self._Hz = Hamiltonian(Hz, dim=self._system.dimension)
+                self._Hz = Hamiltonian(Hz, dim=self._system.dimension)
+            else:
+                extra_terms = []
+                # Add zeeman terms only if there is a field present
+                if not np.array_equal(self._B, [0, 0, 0]):
```

Comment:
The non celio case just above it results in a matrix of zeros which is added to the system hamiltonian once. While the celio case would still work without the check, but due to the nature of the method it would result in extra for loop iterations during the evolve for each term added which for performance we don't want.